graph/pull.go

```diff
@@ -457,6 +463,9 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
 		}
 		downloads[i].img = img
 
+		layers = append(layers, img.ID)
+		s.graph.Retain([]string{img.ID}, manifest.Name)
```

Comment:
There's a deferred call to `Retain` above but I think it makes the mistake of using the current value of `layers`. In this `for` loop `layers = append(layers, img.ID)` changes the value of `layers` so the defered `Unretain` will always be called with the initial empty slice.

We could update the above `defer` to use a closure so that the `layers` isn't evaluated until the deferred call is made:

```go
layers := []string{}
defer func() {
	s.graph.Unretain(layers, manifest.Name)
}()
```